File: src/main/java/io/github/dqualizer/dqlang/types/dam/Identifiable.java
package io.github.dqualizer.dqlang.types.dam;

import lombok.experimental.SuperBuilder;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.Objects;


public abstract class Identifiable {
    @Id
    private String id;

    /**
     * Read the class information stored in Mongo
     */
    private String _class;

    public final String getId() {
        return id;
    }

    public final String getSerializationClass() {
        return _class;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Identifiable that)) {
            return false;
        }
        if (id == null) {
            return false;
        }

        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return id != null ? id.hashCode() : System.identityHashCode(this);
    }
}

File: src/main/kotlin/io/github/dqualizer/dqlang/types/dam/Identifiable.kt
package io.github.dqualizer.dqlang.types.dam

import org.springframework.data.annotation.Id

abstract class Identifiable {
    @Id
    var id: String? = null

    /**
     * Read the class information stored in Mongo
     */
    val _class: String? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is Identifiable) {
            return false
        }
        return if (id == null) {
            false
        } else id == other.id
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: System.identityHashCode(this)
    }
}

